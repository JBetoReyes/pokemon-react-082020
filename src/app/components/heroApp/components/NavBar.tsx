/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { useContext } from 'react';
import {
  Link,
  NavLink,
  RouteComponentProps,
  useHistory,
} from 'react-router-dom';
import { AppClickEvent } from '@typings/htmlEvents';
import AppContext, { IAppContext } from '../app.context';
import { logOut } from '../auth/auth.actions';

type Props = Pick<RouteComponentProps, 'history'>;

export default (): JSX.Element => {
  const { user, setUser } = useContext(AppContext) as IAppContext;
  const history = useHistory();
  const { name } = user;
  const handleLogout = (e: AppClickEvent) => {
    setUser(logOut());
    history.replace('/login');
  };
  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <Link className="navbar-brand" to="/">
        Asociaciones
      </Link>

      <div className="navbar-collapse">
        <div className="navbar-nav">
          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/dc"
          >
            DC
          </NavLink>
          <NavLink
            activeClassName="active"
            className="nav-item nav-link"
            exact
            to="/search"
          >
            Search
          </NavLink>
        </div>
      </div>

      <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
        <ul className="navbar-nav ml-auto">
          <span className="nav-item nav-link text-info">{name}</span>
          <button
            type="button"
            className="nav-item nav-link btn"
            onClick={handleLogout}
          >
            Logout
          </button>
        </ul>
      </div>
    </nav>
  );
};
