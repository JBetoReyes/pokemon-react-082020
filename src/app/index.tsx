import React from 'react';
import ReactDOM from 'react-dom';
// import Home from './home/Home.component';
import App from './containers/App.component';
import '@styles/_main.scss';

ReactDOM.render(<App /> ,document.getElementById('app'));
