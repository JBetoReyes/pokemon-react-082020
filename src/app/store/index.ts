import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { IPokemon } from '../models/pokemonModel';

import rootReducer from './reducers';

const composeEnhancers =
  (typeof window !== 'undefined' &&
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;
const pokemons: IPokemon[] = localStorage.getItem('pokemons')
  ? JSON.parse(localStorage.getItem('pokemons') as string)
  : [];
export default createStore(
  rootReducer,
  {
    exploreCarousel: {
      isloading: true,
      pokemons,
    },
  },
  composeEnhancers(applyMiddleware(thunk))
);
