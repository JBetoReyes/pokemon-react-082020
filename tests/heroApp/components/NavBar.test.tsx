/* eslint-disable @typescript-eslint/no-explicit-any */
import AppContext from '@components/heroApp/app.context';
import { IAppContext } from '@components/heroApp/app.model';
import NavBar from '@components/heroApp/components/NavBar';
import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Router } from 'react-router-dom';

describe('Navbar component tests', () => {
  let userContext: IAppContext = {
    setUser: jest.fn(),
    user: { logged: true, name: 'test' },
  };
  let historyMock = {
    push: jest.fn(),
    location: {},
    replace: jest.fn(),
    listen: jest.fn(),
    createHref: jest.fn(),
  };
  let wrapper = mount(
    <AppContext.Provider value={userContext}>
      <MemoryRouter>
        <Router history={historyMock as any}>
          <NavBar />
        </Router>
      </MemoryRouter>
    </AppContext.Provider>
  );
  beforeEach(() => {
    userContext = {
      setUser: jest.fn(),
      user: { logged: true, name: 'test' },
    };
    historyMock = {
      push: jest.fn(),
      location: {},
      replace: jest.fn(),
      listen: jest.fn(),
      createHref: jest.fn(),
    };
    wrapper = mount(
      <AppContext.Provider value={userContext}>
        <MemoryRouter>
          <Router history={historyMock as any}>
            <NavBar />
          </Router>
        </MemoryRouter>
      </AppContext.Provider>
    );
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test('Should match with the snapshot', () => {
    expect(wrapper.find('nav.navbar')).toMatchSnapshot();
  });
  test('Should clean the user and navigates to the login page when loging out', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    wrapper.find('button').prop('onClick')({} as any);
    expect(userContext.setUser).toHaveBeenCalledTimes(1);
    expect(userContext.setUser).toHaveBeenCalledWith({ type: '[auth] LOGOUT' });
    expect(historyMock.replace).toHaveBeenCalled();
    expect(historyMock.replace).toHaveBeenCalledWith('/login');
  });
});
