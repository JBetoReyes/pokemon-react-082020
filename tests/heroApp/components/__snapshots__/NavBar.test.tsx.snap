// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component tests Should match with the snapshot 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "p40pm2",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "p40pm2",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Router
      history={
        Object {
          "createHref": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": null,
                },
              ],
              Array [
                Object {
                  "hash": "",
                  "pathname": "/marvel",
                  "search": "",
                  "state": null,
                },
              ],
              Array [
                Object {
                  "hash": "",
                  "pathname": "/dc",
                  "search": "",
                  "state": null,
                },
              ],
              Array [
                Object {
                  "hash": "",
                  "pathname": "/search",
                  "search": "",
                  "state": null,
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "listen": [MockFunction] {
            "calls": Array [
              Array [
                [Function],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          },
          "location": Object {},
          "push": [MockFunction],
          "replace": [MockFunction],
        }
      }
    >
      <Component>
        <nav
          className="navbar navbar-expand-sm navbar-dark bg-dark"
        >
          <Link
            className="navbar-brand"
            to="/"
          >
            <LinkAnchor
              className="navbar-brand"
              navigate={[Function]}
            >
              <a
                className="navbar-brand"
                onClick={[Function]}
              >
                Asociaciones
              </a>
            </LinkAnchor>
          </Link>
          <div
            className="navbar-collapse"
          >
            <div
              className="navbar-nav"
            >
              <NavLink
                activeClassName="active"
                className="nav-item nav-link"
                exact={true}
                to="/marvel"
              >
                <Link
                  aria-current={null}
                  className="nav-item nav-link"
                  to={
                    Object {
                      "hash": "",
                      "pathname": "/marvel",
                      "search": "",
                      "state": null,
                    }
                  }
                >
                  <LinkAnchor
                    aria-current={null}
                    className="nav-item nav-link"
                    navigate={[Function]}
                  >
                    <a
                      aria-current={null}
                      className="nav-item nav-link"
                      onClick={[Function]}
                    >
                      Marvel
                    </a>
                  </LinkAnchor>
                </Link>
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-item nav-link"
                exact={true}
                to="/dc"
              >
                <Link
                  aria-current={null}
                  className="nav-item nav-link"
                  to={
                    Object {
                      "hash": "",
                      "pathname": "/dc",
                      "search": "",
                      "state": null,
                    }
                  }
                >
                  <LinkAnchor
                    aria-current={null}
                    className="nav-item nav-link"
                    navigate={[Function]}
                  >
                    <a
                      aria-current={null}
                      className="nav-item nav-link"
                      onClick={[Function]}
                    >
                      DC
                    </a>
                  </LinkAnchor>
                </Link>
              </NavLink>
              <NavLink
                activeClassName="active"
                className="nav-item nav-link"
                exact={true}
                to="/search"
              >
                <Link
                  aria-current={null}
                  className="nav-item nav-link"
                  to={
                    Object {
                      "hash": "",
                      "pathname": "/search",
                      "search": "",
                      "state": null,
                    }
                  }
                >
                  <LinkAnchor
                    aria-current={null}
                    className="nav-item nav-link"
                    navigate={[Function]}
                  >
                    <a
                      aria-current={null}
                      className="nav-item nav-link"
                      onClick={[Function]}
                    >
                      Search
                    </a>
                  </LinkAnchor>
                </Link>
              </NavLink>
            </div>
          </div>
          <div
            className="navbar-collapse collapse w-100 order-3 dual-collapse2"
          >
            <ul
              className="navbar-nav ml-auto"
            >
              <span
                className="nav-item nav-link text-info"
              >
                test
              </span>
              <button
                className="nav-item nav-link btn"
                onClick={[Function]}
                type="button"
              >
                Logout
              </button>
            </ul>
          </div>
        </nav>
      </Component>
    </Router>
  </Router>
</MemoryRouter>
`;
